### PR TITLE
Issue24 copy on interception

### PR DESCRIPTION
**Pull request by optinirr**

### CODE REVIEW

File: playback/tape_recorder.py [hunk 1]
```diff
@@ -873,16 +869,20 @@ def _output_interception_key(alias, invocation_number):
 
 
 class RecordingParameters(object):
-    def __init__(self, sampling_rate=1.0, ignore_enforced_sampling=False, skipped=False):
+    def __init__(self, sampling_rate=1.0, ignore_enforced_sampling=False,
+                 skipped=False, copy_data_on_intercepion=False):
         """
-        :param sampling_rate: Sampling rate for recording, between 0 and 1. Default is 0
+        :param recording_parameters: Optional sampling rate (between 0 and 1) to applied on recording. Default is 1
```
Comment (eitany) [inline playback/tape_recorder.py:875]: Fix

File: playback/tape_recorder.py [hunk 2]
```diff
@@ -873,16 +869,20 @@ def _output_interception_key(alias, invocation_number):
...
         :type sampling_rate: float
-        :param skipped: Recording should be skipped all together
+        :param ignore_enforced_sampling: Whether to ignore enforce sample explicit calls
+        :type ignore_enforced_sampling: bool
+        :param skipped: This class should be skipped for recording
```
Comment (eitany) [inline playback/tape_recorder.py:879]: Missing docstring for the new flag

File: tests/test_tape_recorder.py [hunk 3]
```diff
@@ -1065,3 +1081,27 @@ def execute(self):
                                                       playback_function=lambda recording: Operation().execute())
         self._assert_playback_vs_recording(playback_result, result)
         wrapped.assert_called()
+
+    def test_record_and_playback_basic_operation_data_interception_copy_data(self):
```
Comment (eitany) [inline tests/test_tape_recorder.py:1085]: You are not checking in the recording the data is a copy and not original input
You should fetch last recording after execute and do recording.get_data and see you got a copy
So you need to change this from integer input to a dict input do you can actually check if it is the same instance

**Review by eitany — APPROVED** [review]
This will require a bump in version and update in sdk/euclid due to RecordingParameters change

File: playback/tape_recorder.py [hunk 4]
```diff
@@ -752,11 +747,12 @@ def _execute_func_and_record_interception(self, func, interception_key, args, kw
...
                 return result
 
-            try:
-                recorded_result = copy(recorded_result)  # prevent changes in some fields of input to break recording.
-            except Exception as ex:
-                _logger.warning(u"recorded data couldn't be copied (type={} exception={})".format(
-                    type(recorded_result), repr(ex)))
+            if self._active_recording_parameters.copy_data_on_intercepion:
+                try:
+                    recorded_result = decode(encode(recorded_result))
```
[the remark this answers is not in the document]
Comment (optinirr) [reply playback/tape_recorder.py:752]: whats the reason for that?

[the remark this answers is not in the document]
Comment (eitany) [reply playback/tape_recorder.py:752]: For using True and not False or for passing it explicitly?

File: playback/tape_recorder.py [hunk 5]
```diff
@@ -279,21 +278,16 @@ def operation(self, metadata_extractor=None):
...
+    def recording_params(self, recording_parameters=None, **kwargs):
         """
-        :param sampling_rate: Optional sampling rate (between 0 and 1) to applied on recording. Default is 1
-        :type sampling_rate: float
-        :param ignore_enforced_sampling: Whether to ignore enforce sample explicit calls
-        :type ignore_enforced_sampling: bool
-        :param skipped: This class should be skipped for recording
-        :type skipped: bool
+        :param recording_parameters: Recording Parameters objects for this operation
+        :type sampling_rate: RecordingParameters
```
Comment (eitany) [inline playback/tape_recorder.py:284]: fix

File: tests/test_tape_recorder.py (hunk 3 above)
Comment (optinirr) [reply tests/test_tape_recorder.py]: the in memory save recording encodes it anyway, so it kind passes in any event. I can add a patch that checks that encode is called with 5.

File: playback/tape_recorder.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (optinirr) [reply playback/tape_recorder.py:752]: the default is true

[the remark this answers is not in the document]
Comment (optinirr) [reply playback/tape_recorder.py:752]: but I changed it.

**Pull request closed without merging**